### ISSUE TITLE
error message "Directory not empty" in mv, is confusion

### ISSUE THREAD

**Issue by niron1** [uutils/coreutils#5102]
in file **coreutils/src/uu/mv/src/mv.rs:488**
when you try to move a directory to a location already containing a directory with the same name it would just write "Directory not empty"
first, this is technically a wrong error message because there is no requirement that the destination would be empty. 
the destination might as well be populated with some content, we just require that it would not contain a directory with the same name.
also, the error message is confusing because it doesn't state that the problem is with the destination. One can think that the problem is actually with the source, and that the source directory having some kind of attribute that would require it to be empty prior to moving.
I would suggest to change the error message to: "a directory with the same name already exists at destination"

**Comment by sylvestre** [comment]
could you please provide an example how to reproduce this ?

**Comment by niron1** [comment]
sure:

$:rm -rf /tmp/node_modules
$:mkdir -p proj1/node_modules
$:touch proj1/node_modules/foo
$:mkdir -p proj2/node_modules
$:touch proj2/node_modules/bar
$:mv proj1/node_modules /tmp
$:mv proj2/node_modules /tmp
mv: cannot move 'proj2/node_modules' to '/tmp/node_modules': Directory not empty

**Comment by sylvestre** [comment]
so, if I understand correctly, this issue also happen on the GNU side?
thanks

**Comment by xuky76** [comment]
The issue is that the directory is non-empty and you are attempting to overwrite data.

`mkdir -p foo/bar bar && mv bar foo` will execute just fine
`mkdir -p foo/bar bar && touch foo/bar/ram && mv bar foo` will not

**Comment by tertsdiepraam** [comment]
> I'm not that much familiar with the separation of responsibility among linux packages. I know that linux' utils are based on GNU, in general, I was looking for the source code of whomever maintains the 'mv' utility,

The utils that we maintain here are probably not on your system. That would most likely be the GNU utils. GNU does not operate on GitHub, but they have mailing lists where you can submit suggestions: https://www.gnu.org/software/coreutils/coreutils.html#mailinglists

Our goal is to match the GNU behaviour, so if you get GNU to change this behaviour, we'll follow suit :)

**Comment by niron1** [comment]
> > I'm not that much familiar with the separation of responsibility among linux packages. I know that linux' utils are based on GNU, in general, I was looking for the source code of whomever maintains the 'mv' utility,
> 
> The utils that we maintain here are probably not on your system. That would most likely be the GNU utils. GNU does not operate on GitHub, but they have mailing lists where you can submit suggestions: https://www.gnu.org/software/coreutils/coreutils.html#mailinglists
> 
> Our goal is to match the GNU behaviour, so if you get GNU to change this behaviour, we'll follow suit :)

yes, you are right, it's GNU:
mv --help
.
.
GNU coreutils online help: <https://www.gnu.org/software/coreutils/>
Full documentation <https://www.gnu.org/software/coreutils/mv>
or available locally via: info '(coreutils) mv invocation'

**Comment by niron1** [comment]
> The issue is that the directory is non-empty and you are attempting to overwrite data.
> 
> `mkdir -p foo/bar bar && mv bar foo` will execute just fine `mkdir -p foo/bar bar && touch foo/bar/ram && mv bar foo` will not

my issue was about improving the phrasing of the error message

**Comment by xuky76** [comment]
> > The issue is that the directory is non-empty and you are attempting to overwrite data.
> > `mkdir -p foo/bar bar && mv bar foo` will execute just fine `mkdir -p foo/bar bar && touch foo/bar/ram && mv bar foo` will not
> 
> my issue was about improving the phrasing of the error message

Yeah I understand the confusion but it might just be your understanding of how mv works. There can be a destination directory that shares the same name - It just can't have anything in it, hence their error.
i.e.
`mkdir -p foo/bar bar && touch bar/foo && mv bar foo` 
VS
`mkdir -p foo/bar bar && touch foo/bar/ram && mv bar foo`

**Comment by niron1** [comment]
> > > The issue is that the directory is non-empty and you are attempting to overwrite data.
> > > `mkdir -p foo/bar bar && mv bar foo` will execute just fine `mkdir -p foo/bar bar && touch foo/bar/ram && mv bar foo` will not
> > 
> > 
> > my issue was about improving the phrasing of the error message
> 
> Yeah I understand the confusion but it might just be your understanding of how mv works. There can be a destination directory that shares the same name - It just can't have anything in it, hence their error. i.e. `mkdir -p foo/bar bar && touch bar/foo && mv bar foo` VS `mkdir -p foo/bar bar && touch foo/bar/ram && mv bar foo`

exactly. your second example would trigger the suboptimal error message

**Comment by tertsdiepraam** [comment]
@niron1 do you have a link to the mailing list message? I'd love to follow the discussion there.

**Comment by niron1** [comment]
I guess not. I initially opened the bug by sending an email, then I
responded to subsequent emails using "reply all"....

**Comment by tertsdiepraam** [comment]
Found it :) https://lists.gnu.org/archive/html/bug-coreutils/2023-07/msg00030.html

**Comment by PThorpe92** [comment]
I like that solution. Changing

    Cannot move {} into {}: Directory not empty.
into

    Cannot overwrite: A non-empty directory: '{}' exists at destination 
Or something similar.

**Comment by aaron-ang** [comment]
seems like GNU's default message: `Cannot overwrite {target}: Directory not empty` is clear enough. Should we revert to that?

**Comment by aaron-ang** [comment]
@sylvestre wdyt?

**Comment by HackingRepo** [comment]
it fixed, now @sylvestre 

```css
relunsec@relunsec:~/software/coreutils$ mv docs /tmp
relunsec@relunsec:~/software/coreutils$ ls
CODE_OF_CONDUCT.md  Cargo.lock  Cross.toml      GNUmakefile  Makefile   README.package.md  Test  build.rs   flake.lock  fuzz         renovate.json  target  util
CONTRIBUTING.md     Cargo.toml  DEVELOPMENT.md  LICENSE      README.md  SECURITY.md        a     deny.toml  flake.nix   oranda.json  src            tests
relunsec@relunsec:~/software/coreutils$ 
```

no more directory not empty confusing messages, it seems

**Comment by aaron-ang** [comment]
hi @HackingRepo, your example is wrong. It does not output a message in your case because `/tmp` does not contain `docs` dir. If you rerun `mv docs /tmp`now with a non-empty `/tmp/docs`, you should see the message.

**Comment by HackingRepo** [comment]
ok

```css
relunsec@relunsec:~/software/coreutils$ target/debug/mv Test /tmp
mv: cannot move 'Test' to '/tmp/Test': Directory not empty
relunsec@relunsec:~/software/coreutils$ 
```

i see that

however gnu too

```css
relunsec@relunsec:~/software/coreutils$ gnumv Test /tmp
gnumv: inter-device move failed: 'Test' to '/tmp/Test'; unable to remove target: Directory not empty
relunsec@relunsec:~/software/coreutils$ 
```

say directory not empty

**Comment by aaron-ang** [comment]
@HackingRepo yep, that's the reason for this issue. The author believes that the warning message is confusing and doesn't match GNU.

**Comment by HackingRepo** [comment]
but not highly confusing, however Directory not empty is a correct error message, just gnu verbose
